stepflow-rs/crates/stepflow-server/tests/tracing_integration_tests.rs

```diff
@@ -0,0 +1,462 @@
+        .step(
+            StepBuilder::builtin_step("step1", "create_messages")
+                .input_literal(json!({
+                    "user_prompt": "Step 1"
+                }))
+                .build(),
+        )
+        .step(
+            StepBuilder::builtin_step("step2", "create_messages")
+                .input_literal(json!({
+                    "user_prompt": "Step 2"
+                }))
+                .build(),
+        )
+        .step(
+            StepBuilder::builtin_step("step3", "create_messages")
+                .input_literal(json!({
+                    "user_prompt": "Step 3"
+                }))
+                .build(),
+        )
+        .build();
+
+    // Store the workflow first
+    let store_request = Request::builder()
+        .uri("/api/v1/flows")
+        .method("POST")
+        .header("content-type", "application/json")
+        .body(Body::from(
+            serde_json::to_string(&json!({
+                "flow": workflow
+            }))
+            .unwrap(),
+        ))
+        .unwrap();
+
+    let response = app.clone().oneshot(store_request).await.unwrap();
+    assert_eq!(response.status(), StatusCode::OK);
+
+    let body = axum::body::to_bytes(response.into_body(), usize::MAX)
+        .await
+        .unwrap();
+    let store_response: serde_json::Value = serde_json::from_slice(&body).unwrap();
+    let flow_id = store_response["flowId"].as_str().unwrap();
+
+    // Execute workflow via API
+    let execute_request = Request::builder()
+        .uri("/api/v1/runs")
+        .method("POST")
+        .header("content-type", "application/json")
+        .body(Body::from(
+            serde_json::to_string(&json!({
+                "flowId": flow_id,
+                "input": {},
+                "debug": false
+            }))
+            .unwrap(),
+        ))
+        .unwrap();
+
+    let response = app.oneshot(execute_request).await.unwrap();
+    assert_eq!(response.status(), StatusCode::OK);
+
+    // Wait for traces to be exported
+    println!("⏳ Waiting for traces to export (bidirectional workflow)...");
+    // Don't call fastrace::flush() directly - it can block the tokio runtime
+    // Instead, just wait for the background exporter to send traces
+    tokio::time::sleep(tokio::time::Duration::from_secs(3)).await;
+
+    // Verify traces were collected
+    let trace_file = get_trace_file();
+    println!("📁 Reading traces from: {}", trace_file.display());
+    let traces = read_traces(&trace_file);
+
+    if traces.is_empty() {
+        println!(
+            "⚠️  No traces collected - trace file exists: {}",
+            trace_file.exists()
+        );
+        if trace_file.exists() {
+            println!(
+                "   File size: {} bytes",
+                std::fs::metadata(&trace_file).unwrap().len()
+            );
+        }
+        return;
+    }
+
+    // Look for workflow execution span
+    let workflow_span = find_span(&traces, "flow_execution");
```

Comment:
Should also look for nested spans corresponding to the specific flow ID, and the bidirectional calls.

Reply:
Even better would be to see if we could use the in-memory collector for these tests (although, that may be difficult since we need to connect to the execution layer).

Either way, we should modify these to:

a) Capture the run_id from the execution request, and use that to search for traces.
b) Capture the trace and sub-traces
c) Then look for more specific spans (including bidirectional calls)

This should help make sure we properly wire all the things.